crates/wasmparser/src/readers/core/memories.rs

```diff
@@ -32,23 +32,40 @@ impl<'a> FromReader<'a> for MemoryType {
         let has_max = flags & 0b0001 != 0;
         let has_page_size = flags & 0b1000 != 0;
 
+        if memory64 && !reader.memory64() {
+            bail!(pos, "memory64 must be enabled for 64-bit memories");
+        }
+        if shared && !reader.threads() {
+            bail!(pos, "threads must be enabled for shared memories");
+        }
+        if has_page_size && !reader.custom_page_sizes() {
+            bail!(
+                pos,
+                "the custom page sizes proposal must be enabled to customize a memory's page size"
+            );
+        }
+
         Ok(MemoryType {
             memory64,
             shared,
-            initial: if memory64 {
+            initial: if reader.memory64() {
                 reader.read_var_u64()?
             } else {
                 reader.read_var_u32()?.into()
             },
             maximum: if !has_max {
                 None
-            } else if memory64 {
+            } else if reader.memory64() {
                 Some(reader.read_var_u64()?)
             } else {
                 Some(reader.read_var_u32()?.into())
             },
             page_size_log2: if has_page_size {
-                Some(reader.read_var_u32()?)
+                let val = reader.read_var_u32()?;
+                if val >= 64 {
+                    bail!(pos, "invalid custom page size");
+                }
```

Comment:
At least with the [current spec](https://github.com/WebAssembly/custom-page-sizes/blob/main/proposals/custom-page-sizes/Overview.md) I think that it's a valid binary to specify an arbitrary u32 here, but only validation-invalid if it's >= 64. Does this cause problems with the preexisting upstream spec tests though?

Reply:
Yeah, this is required by an upstream test: https://github.com/WebAssembly/custom-page-sizes/blob/edb9c5b5/test/core/custom-page-sizes/custom-page-sizes-invalid.wast#L90 .

I think it needs to be malformed because otherwise there would be no way to represent bigger values in the text format (the text format has the $2^n$ page size and the binary format just stores $n$).